sklearn/semi_supervised/tests/test_self_training.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_raise_message
+from sklearn.utils.testing import assert_warns_message
+from sklearn.exceptions import NotFittedError
+
+from sklearn.semi_supervised import SelfTrainingClassifier
+from sklearn.dummy import DummyClassifier
+from sklearn.model_selection import ParameterGrid
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.svm import SVC
+from sklearn.model_selection import train_test_split
+from sklearn.datasets import load_iris
+from sklearn.utils import check_random_state
+from sklearn.preprocessing import LabelBinarizer
+
+# Author: Oliver Rausch <rauscho@ethz.ch>
+# License: BSD 3 clause
+
+
+rng = check_random_state(0)
+
+# load the iris dataset and randomly permute it
+iris = load_iris()
+perm = rng.permutation(iris.target.size)
+iris.data = iris.data[perm]
+iris.target = iris.target[perm]
+X_train, X_test, y_train, y_test = train_test_split(iris.data,
+                                                    iris.target,
+                                                    random_state=rng)
+
+limit = 50
+
+y_train_missing_labels = y_train.copy()
+y_train_missing_labels[limit:] = -1
+y_train_missing_labels_strings = y_train_missing_labels.copy()
+mapping = {0: 'A', 1: 'B', 2: 'C', -1: -1}
+y_train_missing_strings = np.vectorize(mapping.get)(y_train_missing_labels)
+
+lb = LabelBinarizer()
+y_train_missing_dummies = lb.fit_transform(y_train_missing_labels)
+
+
+def test_classification():
+    # Check classification for various parameter settings.
+    # Also assert that predictions for strings and numerical labels are equal.
+    # Also test for multioutput classification
+    grid = ParameterGrid({"max_iter": [1, 50, 100],
+                          "threshold": [0.0, 0.5, 0.9]})
+
+    for base_classifier in [DummyClassifier(random_state=0),
+                            DecisionTreeClassifier(random_state=0),
+                            KNeighborsClassifier(),
+                            SVC(gamma="scale", probability=True,
+                                random_state=0)]:
+        for params in grid:
+            st = SelfTrainingClassifier(base_classifier, **params)
+            st_string = SelfTrainingClassifier(base_classifier, **params)
```

Comment:
personal taste: define `st_string` right before it's used on line 64, or skip a line after this one to have clearly defined blocks